Gillian's additions to CSE's OCaml library.

GillianCore/cse is CSE's lib/ (hand-written) and lib/extracted (generated by
extraction), plus exactly this patch. Every hunk is an addition:

  * none, empty and loc, the three GIL values with no counterpart in CSE's
    verified value language, as constructors of the SMT Val datatype, with
    their type tests and coercions (extracted/, syntax/, smt/smt.ml);
  * `open Extraction_utils`, where Gillian's build renames CSE's utils library.

Generated by `scripts/vendor-cse/vendor.sh --update-patch`; do not edit by hand.
Check it with `scripts/vendor-cse/vendor.sh --check`.

--- a/extracted/extracted.ml
+++ b/extracted/extracted.ml
@@ -3393,6 +3393,9 @@
 
 type preval =
 | PVNull
+| PVGillianNone
+| PVGillianEmpty
+| PVGillianLoc of int
 | PVBool of bool
 | PVNat of int
 | PVRat of qp
@@ -3403,6 +3406,9 @@
 type type0 =
 | TVal
 | TNull
+| TGillianNone
+| TGillianEmpty
+| TGillianLoc
 | TBool
 | TNat
 | TRat
@@ -4721,6 +4727,9 @@
 let rec encode_type = function
 | TVal -> _UU03c3__val
 | TNull -> _UU03c3__null
+| TGillianNone -> _UU03c3__val
+| TGillianEmpty -> _UU03c3__val
+| TGillianLoc -> _UU03c3__val
 | TBool -> _UU03c3__bool
 | TNat -> _UU03c3__int
 | TRat -> _UU03c3__real
@@ -4806,6 +4815,21 @@
 let c_null_val =
   IdSimple ('n'::('u'::('l'::('l'::('V'::('a'::('l'::[])))))))
 
+(** val c_gillian_none_val : identifier **)
+
+let c_gillian_none_val =
+  IdSimple ('n'::('o'::('n'::('e'::[]))))
+
+(** val c_gillian_empty_val : identifier **)
+
+let c_gillian_empty_val =
+  IdSimple ('e'::('m'::('p'::('t'::('y'::[])))))
+
+(** val c_gillian_loc_val : identifier **)
+
+let c_gillian_loc_val =
+  IdSimple ('l'::('o'::('c'::[])))
+
 (** val c_bool_val : identifier **)
 
 let c_bool_val =
@@ -4845,7 +4869,22 @@
 
 let null_val =
   null_val_of null
+
+(** val gillian_none_val : term **)
+
+let gillian_none_val =
+  TApp (c_gillian_none_val, None, [])
+
+(** val gillian_empty_val : term **)
+
+let gillian_empty_val =
+  TApp (c_gillian_empty_val, None, [])
+
+(** val gillian_loc_val : term -> term **)
 
+let gillian_loc_val t =
+  TApp (c_gillian_loc_val, None, (t :: []))
+
 (** val bool_val : term -> term **)
 
 let bool_val t =
@@ -4927,6 +4966,11 @@
 let g_null_val =
   IdSimple ('g'::('e'::('t'::('N'::('u'::('l'::('l'::[])))))))
 
+(** val g_gillian_loc_val : identifier **)
+
+let g_gillian_loc_val =
+  IdSimple ('g'::('e'::('t'::('L'::('o'::('c'::[]))))))
+
 (** val g_bool_val : identifier **)
 
 let g_bool_val =
@@ -5007,6 +5051,21 @@
 let p_null_val =
   identifier_add_prefix ('i'::('s'::('-'::[]))) c_null_val
 
+(** val p_gillian_none_val : identifier **)
+
+let p_gillian_none_val =
+  identifier_add_prefix ('i'::('s'::('-'::[]))) c_gillian_none_val
+
+(** val p_gillian_empty_val : identifier **)
+
+let p_gillian_empty_val =
+  identifier_add_prefix ('i'::('s'::('-'::[]))) c_gillian_empty_val
+
+(** val p_gillian_loc_val : identifier **)
+
+let p_gillian_loc_val =
+  identifier_add_prefix ('i'::('s'::('-'::[]))) c_gillian_loc_val
+
 (** val p_bool_val : identifier **)
 
 let p_bool_val =
@@ -5042,6 +5101,21 @@
 let is_null_val t =
   TApp (p_null_val, None, (t :: []))
 
+(** val is_gillian_none_val : term -> term **)
+
+let is_gillian_none_val t =
+  TApp (p_gillian_none_val, None, (t :: []))
+
+(** val is_gillian_empty_val : term -> term **)
+
+let is_gillian_empty_val t =
+  TApp (p_gillian_empty_val, None, (t :: []))
+
+(** val is_gillian_loc_val : term -> term **)
+
+let is_gillian_loc_val t =
+  TApp (p_gillian_loc_val, None, (t :: []))
+
 (** val is_bool_val : term -> term **)
 
 let is_bool_val t =
@@ -5120,6 +5194,9 @@
   match _UU03c4_ with
   | TVal -> true_
   | TNull -> is_null_val t
+  | TGillianNone -> is_gillian_none_val t
+  | TGillianEmpty -> is_gillian_empty_val t
+  | TGillianLoc -> is_gillian_loc_val t
   | TBool -> is_bool_val t
   | TNat -> is_nat_val t
   | TRat -> is_rat_val t
@@ -5324,7 +5401,52 @@
               (is_null_val t_some))))))
   in
   lookup0 (gmap_lookup sort_eq_decision sort_countable) _UU03c3_ cases
+
+(** val to_gillian_value :
+    (term -> term) -> ((term * sort) * term gset) -> ((term * sort) * term
+    gset) option **)
+
+let to_gillian_value tester = function
+| (p, _UU03a6_) ->
+  let (t, _UU03c3_) = p in
+  let t_some = get_some_val t in
+  let cases =
+    insert0 (map_insert (gmap_partial_alter sort_eq_decision sort_countable))
+      _UU03c3__val ((t, _UU03c3__val),
+      (union0 (gset_union term_eq_decision term_countable) _UU03a6_
+        (singleton0 (gset_singleton term_eq_decision term_countable)
+          (tester t))))
+      (singletonM0
+        (map_singleton (gmap_partial_alter sort_eq_decision sort_countable)
+          (gmap_empty sort_eq_decision sort_countable)) _UU03c3__maybe_val
+        ((t_some, _UU03c3__val),
+        (union0 (gset_union term_eq_decision term_countable)
+          (union0 (gset_union term_eq_decision term_countable) _UU03a6_
+            (singleton0 (gset_singleton term_eq_decision term_countable)
+              (is_some_val t)))
+          (singleton0 (gset_singleton term_eq_decision term_countable)
+            (tester t_some)))))
+  in
+  lookup0 (gmap_lookup sort_eq_decision sort_countable) _UU03c3_ cases
+
+(** val to_gillian_none :
+    ((term * sort) * term gset) -> ((term * sort) * term gset) option **)
 
+let to_gillian_none =
+  to_gillian_value is_gillian_none_val
+
+(** val to_gillian_empty :
+    ((term * sort) * term gset) -> ((term * sort) * term gset) option **)
+
+let to_gillian_empty =
+  to_gillian_value is_gillian_empty_val
+
+(** val to_gillian_loc :
+    ((term * sort) * term gset) -> ((term * sort) * term gset) option **)
+
+let to_gillian_loc =
+  to_gillian_value is_gillian_loc_val
+
 (** val to_bool :
     ((term * sort) * term gset) -> ((term * sort) * term gset) option **)
 
@@ -5489,6 +5611,9 @@
   match _UU03c4_ with
   | TVal -> to_val ((t, _UU03c3_), _UU03a6_)
   | TNull -> to_null ((t, _UU03c3_), _UU03a6_)
+  | TGillianNone -> to_gillian_none ((t, _UU03c3_), _UU03a6_)
+  | TGillianEmpty -> to_gillian_empty ((t, _UU03c3_), _UU03a6_)
+  | TGillianLoc -> to_gillian_loc ((t, _UU03c3_), _UU03a6_)
   | TBool -> to_bool ((t, _UU03c3_), _UU03a6_)
   | TNat -> to_nat0 ((t, _UU03c3_), _UU03a6_)
   | TRat -> to_rat ((t, _UU03c3_), _UU03a6_)
@@ -5604,6 +5729,22 @@
 let rec encode_preval = function
 | PVNull ->
   Some ((null, _UU03c3__null),
+    (empty0 (gset_empty term_eq_decision term_countable)))
+| PVGillianNone ->
+  Some ((gillian_none_val, _UU03c3__val),
+    (empty0 (gset_empty term_eq_decision term_countable)))
+| PVGillianEmpty ->
+  Some ((gillian_empty_val, _UU03c3__val),
+    (empty0 (gset_empty term_eq_decision term_countable)))
+| PVGillianLoc i ->
+  let rec positive_of_int n0 =
+    if n0 <= 1 then XH
+    else
+      let q = n0 / 2 in
+      if n0 mod 2 = 0 then XO (positive_of_int q) else XI (positive_of_int q)
+  in
+  let z = if i <= 0 then Z0 else Zpos (positive_of_int i) in
+  Some (((gillian_loc_val (int_literal z)), _UU03c3__val),
     (empty0 (gset_empty term_eq_decision term_countable)))
 | PVBool b ->
   Some (((bool_ b), _UU03c3__bool),
--- a/extracted/extracted.mli
+++ b/extracted/extracted.mli
@@ -732,6 +732,9 @@
 
 type preval =
 | PVNull
+| PVGillianNone
+| PVGillianEmpty
+| PVGillianLoc of int
 | PVBool of bool
 | PVNat of int
 | PVRat of qp
@@ -742,6 +745,9 @@
 type type0 =
 | TVal
 | TNull
+| TGillianNone
+| TGillianEmpty
+| TGillianLoc
 | TBool
 | TNat
 | TRat
@@ -1114,6 +1120,12 @@
 
 val c_null_val : identifier
 
+val c_gillian_none_val : identifier
+
+val c_gillian_empty_val : identifier
+
+val c_gillian_loc_val : identifier
+
 val c_bool_val : identifier
 
 val c_nat_val : identifier
@@ -1130,6 +1142,12 @@
 
 val null_val : term
 
+val gillian_none_val : term
+
+val gillian_empty_val : term
+
+val gillian_loc_val : term -> term
+
 val bool_val : term -> term
 
 val nat_val : term -> term
@@ -1152,6 +1170,8 @@
 
 val g_null_val : identifier
 
+val g_gillian_loc_val : identifier
+
 val g_bool_val : identifier
 
 val g_nat_val : identifier
@@ -1183,7 +1203,13 @@
 val get_some_val : term -> term
 
 val p_null_val : identifier
+
+val p_gillian_none_val : identifier
+
+val p_gillian_empty_val : identifier
 
+val p_gillian_loc_val : identifier
+
 val p_bool_val : identifier
 
 val p_nat_val : identifier
@@ -1198,6 +1224,12 @@
 
 val is_null_val : term -> term
 
+val is_gillian_none_val : term -> term
+
+val is_gillian_empty_val : term -> term
+
+val is_gillian_loc_val : term -> term
+
 val is_bool_val : term -> term
 
 val is_nat_val : term -> term
@@ -1245,6 +1277,20 @@
 val to_null :
   ((term * sort) * term gset) -> ((term * sort) * term gset) option
 
+val to_gillian_value :
+  (term -> term) ->
+  ((term * sort) * term gset ->
+  ((term * sort) * term gset) option)
+
+val to_gillian_none :
+  ((term * sort) * term gset) -> ((term * sort) * term gset) option
+
+val to_gillian_empty :
+  ((term * sort) * term gset) -> ((term * sort) * term gset) option
+
+val to_gillian_loc :
+  ((term * sort) * term gset) -> ((term * sort) * term gset) option
+
 val to_bool :
   ((term * sort) * term gset) -> ((term * sort) * term gset) option
 
--- a/smt/smt.ml
+++ b/smt/smt.ml
@@ -1,3 +1,5 @@
+open Extraction_utils
+
 open Syntax
 
 module type S = sig
@@ -68,6 +70,11 @@
               Sexp.List
                 [atom Extracted.c_null_val;
                  Sexp.List [atom Extracted.g_null_val; atom Extracted.s_null]];
+                Sexp.List [atom Extracted.c_gillian_none_val];
+                Sexp.List [atom Extracted.c_gillian_empty_val];
+                Sexp.List
+                  [atom Extracted.c_gillian_loc_val;
+                   Sexp.List [atom Extracted.g_gillian_loc_val; atom Extracted.s_int]];
               Sexp.List
                 [atom Extracted.c_bool_val;
                  Sexp.List [atom Extracted.g_bool_val; atom Extracted.s_bool]];
--- a/smtlib/sort.ml
+++ b/smtlib/sort.ml
@@ -1,3 +1,5 @@
+open Extraction_utils
+
 type t =
   | Param of Sexplib.Sexp.t
   | App of (Sexplib.Sexp.t * t list)
--- a/smtlib/term.ml
+++ b/smtlib/term.ml
@@ -1,3 +1,5 @@
+open Extraction_utils
+
 type t =
   | Var of string
   (* The optional sort is SMT-LIB's [(as f σ)] qualifier: a polymorphic
--- a/syntax/symbexp.ml
+++ b/syntax/symbexp.ml
@@ -1,3 +1,5 @@
+open Extraction_utils
+
 type t =
 | Val of Val.t
 | LVar of string
--- a/syntax/type.ml
+++ b/syntax/type.ml
@@ -1,6 +1,9 @@
 type t =
 | Val
 | Null
+| None
+| Empty
+| Loc
 | Bool
 | Nat
 | Rat
@@ -10,6 +13,9 @@
 let rec to_extracted (t : t) : Extracted.type0 = match t with
   | Val -> Extracted.TVal
   | Null -> Extracted.TNull
+  | None -> Extracted.TGillianNone
+  | Empty -> Extracted.TGillianEmpty
+  | Loc -> Extracted.TGillianLoc
   | Bool -> Extracted.TBool
   | Nat -> Extracted.TNat
   | Rat -> Extracted.TRat
--- a/syntax/type.mli
+++ b/syntax/type.mli
@@ -3,6 +3,9 @@
 type t =
 | Val
 | Null
+| None
+| Empty
+| Loc
 | Bool
 | Nat
 | Rat
--- a/syntax/val.ml
+++ b/syntax/val.ml
@@ -1,5 +1,10 @@
+open Extraction_utils
+
 type t =
 | Null
+| None
+| Empty
+| Loc of int
 | Bool of bool
 | Nat of int
 | Rat of float
@@ -30,6 +35,9 @@
 
 let rec to_extracted (v : t) = match v with
   | Null -> Extracted.PVNull
+  | None -> Extracted.PVGillianNone
+  | Empty -> Extracted.PVGillianEmpty
+  | Loc i -> Extracted.PVGillianLoc i
   | Bool b -> Extracted.PVBool b
   | Nat n -> Extracted.PVNat n
   | Rat r -> Extracted.PVRat (q_of_float r)
--- a/syntax/val.mli
+++ b/syntax/val.mli
@@ -2,6 +2,9 @@
 
 type t =
 | Null
+| None
+| Empty
+| Loc of int
 | Bool of bool
 | Nat of int
 | Rat of float
